File: core/src/ru/myitschool/dcrawler/skills/ArcaneShielding.java
package ru.myitschool.dcrawler.skills;

import com.badlogic.gdx.graphics.Texture;
import com.badlogic.gdx.utils.Array;
import ru.myitschool.dcrawler.effects.ArcaneShieldingEffect;
import ru.myitschool.dcrawler.entities.Entity;
import ru.myitschool.dcrawler.math.DiceAction;
import ru.myitschool.dcrawler.math.MathAction;
import ru.myitschool.dcrawler.skills.action.Action;
import ru.myitschool.dcrawler.skills.action.SimpleAttackAction;
import ru.myitschool.dcrawler.skills.action.SimpleEffectAction;
import ru.myitschool.dcrawler.skills.patterns.CharacterTargetPattern;
import ru.myitschool.dcrawler.skills.patterns.TargetPattern;
import ru.myitschool.dcrawler.skills.targeting.CharacterDisplayer;
import ru.myitschool.dcrawler.skills.targeting.TargetDisplayer;

/**
 * Created by Voyager on 26.11.2017.
 */
public class ArcaneShielding extends Skill {

    MathAction rollAction = new DiceAction(20);

    public ArcaneShielding(Entity doer) {
        super(doer);
        addPlayContainer().getPlayerPlay().addAction(new SimpleEffectAction(this) {
            @Override
            public void successEffect(Target target, int damage, FloatingDamageMark mark) {
                target.getEntity().addEffect(new ArcaneShieldingEffect(getDoer()));
            }
        });
    }

    @Override
    public void maintainDisplayers(Array<TargetDisplayer> displayers) {
        displayers.add(new CharacterDisplayer());
    }

    @Override
    public TargetPattern getPattern() {
        return new CharacterTargetPattern(this);
    }

    @Override
    public Texture getIcon() {
        return new Texture("gods_protection.png");
    }

    @Override
    public String getName() {
        return "Arcane shielding";
    }

    @Override
    public String getDescription() {
        return "This shield will protect its target from any damage for one turn";
    }

    @Override
    public int getSkillAccuracyBonus() {
        return 0;
    }

    @Override
    public int getRange() {
        return 1;
    }

    @Override
    public int getDistanceMin() {
        return 0;
    }

    @Override
    public int getDistanceMax() {
        return 4;
    }

    @Override
    public int getTargetCountMax() {
        return 1;
    }

    @Override
    public int getCooldownMax() {
        return 0;
    }

    @Override
    public int getType() {
        return SKILL_TYPE_ENCOUNTER;
    }

    @Override
    public int getTargetType() {
        return SKILL_TARGET_TYPE_CHARACTER;
    }

    @Override
    public boolean isCheckAllTargets() {
        return false;
    }

    @Override
    public boolean isMarkEverything() {
        return false;
    }

    @Override
    public boolean isMark() {
        return true;
    }

    @Override
    public boolean isObstruct() {
        return false;
    }

    @Override
    public boolean isWallTargets() {
        return false;
    }
}

File: core/src/com/wolg_vlad/dcrawler/entities/skills/ArcaneShielding.kt
package com.wolg_vlad.dcrawler.entities.skills

import com.badlogic.gdx.graphics.Texture
import com.wolg_vlad.dcrawler.effects.ArcaneShieldingEffect
import com.wolg_vlad.dcrawler.entities.Entity
import com.wolg_vlad.dcrawler.entities.skills.action.SimpleEffectAction
import com.wolg_vlad.dcrawler.entities.skills.patterns.CharacterTargetPattern
import com.wolg_vlad.dcrawler.entities.skills.patterns.TargetPattern
import com.wolg_vlad.dcrawler.entities.skills.targeting.CharacterDisplayer
import com.wolg_vlad.dcrawler.entities.skills.targeting.TargetDisplayer
import com.wolg_vlad.dcrawler.entities.skills.targeting.TargetRenderer
import com.wolg_vlad.dcrawler.math.DiceAction
import com.wolg_vlad.dcrawler.math.MathAction

/**
 * Created by Voyager on 26.11.2017.
 */
class ArcaneShielding(doer: Entity) : Skill(doer) {
    override val displayers: Array<TargetDisplayer> = arrayOf(CharacterDisplayer())
    override val pattern: TargetPattern = CharacterTargetPattern(this)
    override val icon: Texture = Texture("gods_protection.png")
    override val name: String = "Arcane shielding"
    override val description: String = "This shield will protect its target from any damage for one turn"
    override val skillAccuracyBonus: Int = 0
    override val range: Int = 1
    override val distanceMin: Int = 0
    override val distanceMax: Int = 4
    override val targetCountMax: Int = 1
    override val cooldownMax: Int = 0
    override val type: Int = SKILL_TYPE_ENCOUNTER
    override val targetType: Int = SKILL_TARGET_TYPE_CHARACTER
    override val isCheckAllTargets: Boolean = false
    override val isMarkEverything: Boolean = false
    override val isMark: Boolean = true
    override val isObstruct: Boolean = false
    override val isWallTargets: Boolean = false

    init {
        addPlayContainer().playerPlay.addAction(object : SimpleEffectAction(this) {
            public override fun successEffect(target: Target, damage: Int, mark: FloatingDamageMark) {
                target.entity!!.addEffect(ArcaneShieldingEffect(target.entity))
            }
        })
    }
}